src/main/java/net/neoforged/neoforge/capabilities/CapabilityRegistry.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.capabilities;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import net.minecraft.resources.ResourceLocation;
+
+/**
+ * Helper class to manage registering capabilities.
+ * You only need this if you are creating your own type (block, entity, item...) of capabilities!
+ *
+ * <p>Look at the source code of {@link BlockCapability}, {@link EntityCapability}, ... for an example.
+ */
+public class CapabilityRegistry<C> {
+    private final Map<ResourceLocation, StoredCap<C>> caps = new HashMap<>();
+    private final CapabilityConstructor<C> constructor;
+
+    public CapabilityRegistry(CapabilityConstructor<C> constructor) {
+        Objects.requireNonNull(constructor);
+        this.constructor = constructor;
+    }
+
+    /**
+     * Creates a new capability with the given name, type class and context class,
+     * or returns an existing one if it was already created.
+     *
+     * @param name         name of the capability
+     * @param typeClass    class of the queried object
+     * @param contextClass class of the additional context
+     * @throws IllegalStateException if a capability with the same name but different type or context class was already created
+     * @throws IllegalStateException if this method is called after {@link RegisterCapabilitiesEvent} is finished posting
+     */
+    public synchronized C create(ResourceLocation name, Class<?> typeClass, Class<?> contextClass) {
+        Objects.requireNonNull(name);
+        Objects.requireNonNull(typeClass);
+        Objects.requireNonNull(contextClass);
+
+        StoredCap<C> ret = caps.get(name);
+        if (ret == null) {
+            if (CapabilityHooks.initFinished)
+                throw new IllegalStateException("Cannot create new capabilities after the register event was posted");
```

Comment:
This is problematic with the current design, because we rely on the `XxxCapability.create` call to happen before `initFinished` is true. In other words, the design relies on classloading... That's not great.

Reply:
Does it really though? That was my initial thought and why I almost left a comment about it, but then I realized that while yes it does rely on classloading, don't you also have to finish registering possible caps to blocks, entities, items, etc during the `RegisterCapabilitiesEvent` or the capability won't be exposed on the thing? At which point you have to reference the capability (which thus causes class loading) before `initFinished` is set to `true` or it won't be an accessible capability.

Reply:
What if you define a capability that you don't actually use? For example `Capabilities.FluidHandler.ENTITY`? It's only registered because it's in the same class as the block and item fluid handler caps.

Reply:
IMO we can just remove this check - people will be able to create caps after the event, but they will just not have any registered provider.

Reply:
Ah right, for querying if something provides it. And I agree, we can just remove the check

Reply:
Alright, removed